### PR TITLE
HIVE-26500: Improve TestHiveMetastore

### PR DESCRIPTION
**Pull request by abstractdog**
### What changes were proposed in this pull request?
Test improvements for TestHiveMetastore, described in jira.

### Why are the changes needed?
Fix flaky test + useful assertion messages.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
Unit tests.

### CODE REVIEW

File: standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/TestHiveMetaStore.java [hunk 1]
```diff
@@ -168,16 +169,16 @@ public void testNameMethods() {
...
-      assertTrue("Spec from name is incorrect", spec.equals(testSpec));
+      assertTrue("Spec from name is incorrect: " + testSpec, spec.equals(testSpec));
 
       List<String> emptyVals = client.partitionNameToVals("");
-      assertTrue("Values should be empty", emptyVals.size() == 0);
+      assertTrue("Values should be empty, instead its size is: " + emptyVals.size(), emptyVals.size() == 0);
 
       Map<String, String> emptySpec =  client.partitionNameToSpec("");
-      assertTrue("Spec should be empty", emptySpec.size() == 0);
+      assertTrue("Spec should be empty, instead its size is: " + emptySpec.size(), emptySpec.size() == 0);
```
Comment (ayushtkn) [inline standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/TestHiveMetaStore.java:181]: Can be changed to ``assertEquals``

File: standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/TestHiveMetaStore.java [hunk 2]
```diff
@@ -3306,22 +3314,38 @@ public void testJDOPersistanceManagerCleanup() throws Exception {
...
-    assertTrue(numObjectsBeforeClose == numObjectsAfterClose);
 
+    MetastoreTestUtils.waitForAssertion("Checking pm cachesize after client close", new Runnable() {
+      @Override
+      public void run() {
+        int numObjectsAfterClose = getJDOPersistanceManagerCacheSize();
+        assertTrue(String.format("numObjectsBeforeUse: %d != numObjectsAfterClose: %d", numObjectsBeforeUse,
+            numObjectsAfterClose), numObjectsBeforeUse == numObjectsAfterClose);
+      }
+    }, 500, 30000);
```
Comment (ayushtkn) [inline standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/TestHiveMetaStore.java:3329]: Can use Lambda & assertTrue can be changed to ``assertEqual``?
Something like:
```
    MetastoreTestUtils.waitForAssertion("Checking pm cachesize after client close", () -> {
      int numObjectsAfterClose = getJDOPersistanceManagerCacheSize();
      assertEquals(String
              .format("numObjectsBeforeUse: %d != numObjectsAfterClose: %d", numObjectsBeforeUse, numObjectsAfterClose),
          numObjectsBeforeUse, numObjectsAfterClose);
    }, 500, 30000);
```

File: standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/TestHiveMetaStore.java [hunk 3]
```diff
@@ -3306,22 +3314,38 @@ public void testJDOPersistanceManagerCleanup() throws Exception {
...
+
+    MetastoreTestUtils.waitForAssertion("Checking pm cachesize after transport close", new Runnable() {
+      @Override
+      public void run() {
+        int numObjectsAfterDroppedConnection = getJDOPersistanceManagerCacheSize();
+        assertTrue(String.format("numObjectsAfterClose: %d != numObjectsAfterDroppedConnection: %d",
+            numObjectsAfterClose, numObjectsAfterDroppedConnection),
+            numObjectsAfterClose == numObjectsAfterDroppedConnection);
+      }
+    }, 500, 30000);
```
Comment (ayushtkn) [inline standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/TestHiveMetaStore.java:3346]: Same as above can use Lambda & change to ``assertEquals``
```
    MetastoreTestUtils.waitForAssertion("Checking pm cachesize after transport close", () -> {
      int numObjectsAfterDroppedConnection = getJDOPersistanceManagerCacheSize();
      assertEquals(String
          .format("numObjectsAfterClose: %d != numObjectsAfterDroppedConnection: %d", numObjectsAfterClose,
              numObjectsAfterDroppedConnection), numObjectsAfterClose, numObjectsAfterDroppedConnection);
    }, 500, 30000);
```

File: standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/utils/MetastoreTestUtils.java [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+   * It can be useful when the assertion doesn't necessarily pass immediately, and it would be hard
+   * to mock into production code in order to wait for some conditions properly. Instead, it just
+   * waits, but not like a constant time before a single attempt (which is easy, but errorprone).
+   * @param assertionContext
+   * @param runnable
+   * @param msBetweenAssertionAttempts
+   * @param msOverallTimeout
+   * @throws Exception
+   */
+  public static void waitForAssertion(String assertionContext, Runnable assertionRunnable, int msBetweenAssertionAttempts,
```
Comment (ayushtkn) [inline standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/utils/MetastoreTestUtils.java:42]: Hadoop common does provide similar util:
https://github.com/apache/hadoop/blob/trunk/hadoop-common-project/hadoop-common/src/test/java/org/apache/hadoop/test/GenericTestUtils.java#L440

It is used in hive-code at a couple of places, but I think there is no hadoop-common test dependency in this module? if it is we can use that? or just give a check if we can copy that if it has something better

File: standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/TestHiveMetaStore.java (hunk 1 above)
Comment (abstractdog) [reply standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/TestHiveMetaStore.java:180]: ack

File: standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/TestHiveMetaStore.java (hunk 2 above)
Comment (abstractdog) [reply standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/TestHiveMetaStore.java]: ack

File: standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/TestHiveMetaStore.java (hunk 3 above)
Comment (abstractdog) [reply standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/TestHiveMetaStore.java]: ack

File: standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/utils/MetastoreTestUtils.java (hunk 4 above)
Comment (abstractdog) [reply standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/utils/MetastoreTestUtils.java]: good point, however, there is no direct dependency on hadoop-common in metastore (only provided, optional, etc.)

I like GenericTestUtils, I implemented pretty much the same, but I think mine is maybe better in a sense that the caller doesn't have to create a boolean function and handle TimeoutException and stuff, instead just write the very same assertion in a runnable which would be originally there without this timeout hack...if the code block throws assertion error, unit test fails automatically as a unit test is expected

Comment (ayushtkn) [reply standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/utils/MetastoreTestUtils.java]: Looks good

**Review by ayushtkn — APPROVED** [review]
LGTM.
Sonar has some complains.
https://sonarcloud.io/project/issues?resolved=false&types=CODE_SMELL&pullRequest=3556&id=apache_hive&open=AYMS1ejLkC-HJGhR5irJ

I am not sure how we are handling them..

**Comment by abstractdog** [comment]
> LGTM. Sonar has some complains. https://sonarcloud.io/project/issues?resolved=false&types=CODE_SMELL&pullRequest=3556&id=apache_hive&open=AYMS1ejLkC-HJGhR5irJ
> 
> I am not sure how we are handling them..

sonar is absolutely right when saying "Using Thread.sleep in a test is just generally a bad idea.", that's exactly what I fixed with this, but it hasn't realized :)

**Pull request merged**